### PR TITLE
use AdditionalUserDataFiles in bootstrap data generation

### LINKED ISSUE #90: Wire up AdditionalUserDataFiles to generation code
/kind feature

**Describe the solution you'd like**
With #67 and #89, we have an `AdditionalUserDataFiles` field, but it's currently not wired up to anything. We need to plumb it through to the bootstrap data generation.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

### PR DESCRIPTION
**Pull request by ykakarap**
**What this PR does / why we need it**:
This PR wires up AdditionalUserDataFiles to generation code

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #90 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note

```

### CODE REVIEW

File: controllers/kubeadmconfig_controller.go [hunk 1]
```diff
@@ -136,10 +136,12 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (ctrl.Result, erro
 			return ctrl.Result{}, err
 		}
 
-		cloudInitData, err := cloudinit.NewInitControlPlane(&cloudinit.ControlPlaneInput{
+		cpi := &cloudinit.ControlPlaneInput{
 			InitConfiguration:    string(initdata),
 			ClusterConfiguration: string(clusterdata),
-		})
+		}
+		cpi.AdditionalFiles = config.Spec.AdditionalUserDataFiles
```
Comment (ykakarap) [inline controllers/kubeadmconfig_controller.go]: Cannot use promoted fields (promoted from BaseUserData) as struct literal. Hence separate line.

File: controllers/kubeadmconfig_controller.go [hunk 2]
```diff
@@ -170,11 +172,13 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (ctrl.Result, erro
...
 	if util.IsControlPlaneMachine(machine) {
 		// TODO return a sensible error if join config is not specified (implies empty configuration)
-		joinData, err := cloudinit.NewJoinControlPlane(&cloudinit.ControlPlaneJoinInput{
+		cpji := &cloudinit.ControlPlaneJoinInput{
 			// TODO do a len check or something here
 			ControlPlaneAddress: fmt.Sprintf("https://%s:%d", cluster.Status.APIEndpoints[0].Host, cluster.Status.APIEndpoints[0].Port),
 			JoinConfiguration:   string(joinBytes),
-		})
+		}
+		cpji.AdditionalFiles = config.Spec.AdditionalUserDataFiles
```
Comment (ykakarap) [inline controllers/kubeadmconfig_controller.go]: Same reason as mentioned above.

File: controllers/kubeadmconfig_controller.go [hunk 3]
```diff
@@ -189,9 +193,11 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (ctrl.Result, erro
 		return ctrl.Result{}, nil
 	}
 
-	joinData, err := cloudinit.NewNode(&cloudinit.NodeInput{
+	ni := &cloudinit.NodeInput{
 		JoinConfiguration: string(joinBytes),
-	})
+	}
+	ni.AdditionalFiles = config.Spec.AdditionalUserDataFiles
```
Comment (ykakarap) [inline controllers/kubeadmconfig_controller.go]: Same reason as mentioned above.

File: controllers/kubeadmconfig_controller.go (hunk 1 above)
Comment (ncdc) [reply controllers/kubeadmconfig_controller.go]: You can do it like this:
```go
cloudInitData, err := cloudinit.NewInitControlPlane(&cloudinit.ControlPlaneInput{
  BaseUserData: cloudinit.BaseUserData{
    AdditionalFiles: config.Spec.AdditionalUserDataFiles,
  },
  InitConfiguration: string(initdata),
  // ...
}
```

Comment (chuckha) [reply controllers/kubeadmconfig_controller.go]: This is an embedded type, from the second paragraph in this section https://golang.org/ref/spec#Struct_types:

> The unqualified type name acts as the field name.


You've just got to use the field name like this:

```
		joinData, err := cloudinit.NewJoinControlPlane(&cloudinit.ControlPlaneJoinInput{
			// TODO do a len check or something here
			ControlPlaneAddress: fmt.Sprintf("https://%s:%d", cluster.Status.APIEndpoints[0].Host, cluster.Status.APIEndpoints[0].Port),
			JoinConfiguration:   string(joinBytes),
			BaseUserData: cloudinit.BaseUserData{
				AdditionalFiles: config.Spec.AdditionalUserDataFiles,
			},
		})
```

Comment (chuckha) [reply controllers/kubeadmconfig_controller.go]: oops. I really need to refresh my tabs more frequently. Apologies for the duplication @ncdc

Comment (ncdc) [reply controllers/kubeadmconfig_controller.go]: It's all good!

Comment (ykakarap) [reply controllers/kubeadmconfig_controller.go]: Thank you for the suggestion @ncdc  and @chuckha 😄 . PR updated.

**Comment by ykakarap** [comment]
working on rebasing.

**Comment by chuckha** [comment]
/approve

thanks so much @ykakarap

**Pull request merged**